File: dpll.py
def dpll(clauses, assignment):
    """
    DPLL algorithm to determine satisfiability of a set of clauses.

    Parameters:
    - clauses: List of lists representing the CNF clauses.
    - assignment: Dictionary representing the current variable assignments.

    Returns:
    - (bool, dict): A tuple where the first element is True if the clauses are satisfiable, and
                    the second element is the assignment that satisfies the clauses.
    """

    # If all clauses are satisfied
    if is_satisfied(clauses, assignment):
        return True, assignment

    # If there is an empty clause
    if any(len(clause) == 0 for clause in clauses):
        return False, {}

    # Perform unit propagation
    clauses, assignment = unit_propagate(clauses, assignment)

    # Perform pure literal assignment
    clauses, assignment = pure_literal_assign(clauses, assignment)

    # If all clauses are satisfied after unit propagation and pure literal assignment
    if is_satisfied(clauses, assignment):
        return True, assignment

    # If there is an empty clause after unit propagation and pure literal assignment
    if any(len(clause) == 0 for clause in clauses):
        return False, {}

    # Choose an unassigned variable
    for clause in clauses:
        for literal in clause:
            var = abs(literal)
            if var not in assignment:
                break
        else:
            continue
        break

    # Recursively check with the variable set to True
    new_assignment = assignment.copy()
    new_assignment[var] = True
    simplified_clauses = [[l for l in c if l != -var] for c in clauses if var not in c]
    result, final_assignment = dpll(simplified_clauses, new_assignment)
    if result:
        return True, final_assignment

    # Recursively check with the variable set to False
    new_assignment = assignment.copy()
    new_assignment[var] = False
    simplified_clauses = [[l for l in c if l != var] for c in clauses if -var not in c]
    return dpll(simplified_clauses, new_assignment)

def is_satisfied(clauses, assignment):
    """Check if all clauses are satisfied with the current assignment."""
    for clause in clauses:
        satisfied = False
        for literal in clause:
            var = abs(literal)
            if var in assignment:
                val = assignment[var]
                if (literal > 0 and val) or (literal < 0 and not val):
                    satisfied = True
                    break
        if not satisfied:
            return False
    return True

def unit_propagate(clauses, assignment):
    """Perform unit propagation."""
    while True:
        unit_clauses = [clause for clause in clauses if len(clause) == 1]
        if not unit_clauses:
            break
        for clause in unit_clauses:
            literal = clause[0]
            var = abs(literal)
            val = literal > 0
            assignment[var] = val
            clauses = [c for c in clauses if literal not in c]
            clauses = [[l for l in c if l != -literal] for c in clauses]
    return clauses, assignment

def pure_literal_assign(clauses, assignment):
    """Assign pure literals."""
    all_literals = [literal for clause in clauses for literal in clause]
    pure_literals = set(literal for literal in all_literals if -literal not in all_literals)
    for literal in pure_literals:
        var = abs(literal)
        val = literal > 0
        assignment[var] = val
        clauses = [c for c in clauses if literal not in c]
        return clauses, assignment

# Test the DPLL algorithm with the given inputs
clauses1 = [[1, 2, 3], [-1, 2, -3], [1, -2, 3], [-1, -2, -3]]
clauses2 = [[1,2,3],[1,2,-3],[1,-2,3],[1,-2,-3],[-1,2,3],[-1,2,-3],[-1,-2,3],[-1,-2,-3]] # [[1, 2, -3], [-1, -2, 3], [-1, -2, -3], [1, 2, 3]]

print("Testing clauses1: ", clauses1)
result1, assignment1 = dpll(clauses1, {})
print("Satisfiable: ", result1)
print("Assignment: ", assignment1)

print("\nTesting clauses2: ", clauses2)
result2, assignment2 = dpll(clauses2, {})
print("Satisfiable: ", result2)
print("Assignment: ", assignment2)

File: dpll/dpll.go
package dpll

import (
	"errors"
	"fmt"
	"maps"
	"math"
	"slices"
	"sync"

	"github.com/Aki0x137/concurrent-sat-solver-go/set"
)

type Literal int
type Clause []Literal
type Formula []Clause

type Assignment map[Literal]bool

type Result struct {
	mu         *sync.RWMutex
	satisfied  bool
	assignment Assignment
}

// checkClauseValidity validates the formula
func checkClauseValidity(formula Formula) bool {
	for _, clause := range formula {
		if len(clause) == 0 {
			return false
		}
	}
	return true
}

// selectLiteral chooses an unassigned literal in the formula. If no literal is found, it return an error
func selectLiteral(formula Formula, assignment Assignment) (Literal, error) {
	for _, clause := range formula {
		for _, literal := range clause {
			absLiteral := math.Abs(float64(literal))
			if _, ok := assignment[Literal(absLiteral)]; !ok {
				return literal, nil
			}
		}
	}
	return 0, errors.New("no literal found")
}

// isSatisified checks if all clauses are satisfied with the current assignment
func isSatisfied(formula Formula, assignment Assignment) bool {
	for _, clause := range formula {
		satisfied := false
		for _, literal := range clause {
			absLiteral := math.Abs(float64(literal))
			if val, ok := assignment[Literal(absLiteral)]; ok {
				if (literal > 0 && val) || (literal < 0 && !val) {
					satisfied = true
					break
				}
			}
		}
		if !satisfied {
			return false
		}
	}

	return true
}

// unitPropagate performs unit propagation on formula, based on curent assignments
func unitPropagate(formula Formula, assignment Assignment) (Formula, Assignment) {
	updatedFormula := slices.Clone(formula)
	updatedAssignment := maps.Clone(assignment)
	for {
		var unitClauses []Clause
		for _, clause := range updatedFormula {
			if len(clause) == 1 {
				unitClauses = append(unitClauses, clause)
			}
		}

		if len(unitClauses) == 0 {
			break
		}

		for _, clause := range unitClauses {
			literal := clause[0]
			absLiteral := math.Abs(float64(literal))
			updatedAssignment[Literal(absLiteral)] = literal > 0

			var filteredFormula Formula
			for _, c := range updatedFormula {
				if !slices.Contains(c, literal) {
					filteredFormula = append(filteredFormula, c)
				}
			}

			var simplifiedFormula Formula
			for _, c := range filteredFormula {
				updatedClause := slices.Clone(c)
				if index := slices.Index(updatedClause, -literal); index >= 0 {
					updatedClause = slices.Delete(updatedClause, index, index+1)
				}
				simplifiedFormula = append(simplifiedFormula, updatedClause)
			}
			updatedFormula = simplifiedFormula
		}
	}

	return updatedFormula, updatedAssignment
}

/*
pureLiteralAssignment checks for pure literals and updates formula.

If a propositional variable occurs with only one polarity in the formula, it is called pure. A pure literal can always be assigned in a way that makes all clauses containing it true. Thus, when it is assigned in such a way, these clauses do not constrain the search anymore, and can be deleted.
*/
func pureLiteralAssignment(formula Formula, assignment Assignment) (Formula, Assignment) {
	updatedFormula := slices.Clone(formula)
	updatedAssignment := maps.Clone(assignment)

	allLiteralsSet := set.NewSet[Literal]()
	for _, clauses := range formula {
		for _, literal := range clauses {
			allLiteralsSet.Add(literal)
		}
	}

	allLiterals := allLiteralsSet.Values()
	pureLiterals := set.NewSet[Literal]()
	for _, literal := range allLiterals {
		if !slices.Contains(allLiterals, -literal) {
			pureLiterals.Add(literal)
		}
	}

	for _, literal := range pureLiterals.Values() {
		absLiteral := math.Abs(float64(literal))
		updatedAssignment[Literal(absLiteral)] = literal > 0

		var filteredFormula Formula
		for _, clause := range updatedFormula {
			if index := slices.Index(clause, literal); index == -1 {
				filteredFormula = append(filteredFormula, clause)
			}
		}
		updatedFormula = filteredFormula
	}

	return updatedFormula, updatedAssignment
}

// simplifyFormula removes truthy clauses and removes redundant literals after an assignment
func simplifyFormula(formula Formula, literal Literal) Formula {
	var simplifiedFormula Formula
	for _, clause := range formula {
		if !slices.Contains(clause, literal) {
			updatedClause := slices.Clone(clause)
			if index := slices.Index(updatedClause, -literal); index >= 0 {
				updatedClause = slices.Delete(updatedClause, index, index+1)
			}
			simplifiedFormula = append(simplifiedFormula, updatedClause)
		}
	}
	return simplifiedFormula
}

func Solve(formula Formula, assignment Assignment) (bool, Assignment) {
	if len(formula) == 0 {
		return false, assignment
	}

	if !checkClauseValidity(formula) {
		return false, assignment
	}

	if isSatisfied(formula, assignment) {
		return true, assignment
	}

	newFormula, newAssignment := unitPropagate(formula, assignment)

	newFormula, newAssignment = pureLiteralAssignment(newFormula, newAssignment)

	if isSatisfied(newFormula, newAssignment) {
		return true, newAssignment
	}

	if !checkClauseValidity(formula) {
		return false, assignment
	}

	selectedLiteral, err := selectLiteral(newFormula, newAssignment)
	if err != nil {
		return false, assignment
	}

	ans := Result{&sync.RWMutex{}, false, Assignment{}}
	var wg sync.WaitGroup

	assignment1 := maps.Clone(newAssignment)
	assignment1[selectedLiteral] = true

	simplifiedFormula := simplifyFormula(newFormula, selectedLiteral)

	wg.Add(1)
	go recursiveSolution(simplifiedFormula, assignment1, &wg, &ans)

	assignment2 := maps.Clone(newAssignment)
	assignment2[selectedLiteral] = false

	simplifiedFormula = simplifyFormula(newFormula, -selectedLiteral)

	wg.Add(1)
	go recursiveSolution(simplifiedFormula, assignment2, &wg, &ans)

	wg.Wait()

	fmt.Println("After wait")

	if ans.satisfied {
		return true, ans.assignment
	} else {
		return false, assignment
	}
}

func recursiveSolution(formula Formula, assignment Assignment, wg *sync.WaitGroup, ans *Result) {
	defer wg.Done()

	fmt.Println("INside recursive", assignment)

	ans.mu.RLock()
	if ans.satisfied {
		return
	}
	ans.mu.RUnlock()

	if len(formula) == 0 {
		return
	}

	if !checkClauseValidity(formula) {
		return
	}

	if isSatisfied(formula, assignment) {
		ans.mu.Lock()
		ans.assignment = assignment
		ans.satisfied = true
		ans.mu.Unlock()
		return
	}

	newFormula, newAssignment := unitPropagate(formula, assignment)

	newFormula, newAssignment = pureLiteralAssignment(newFormula, newAssignment)

	if isSatisfied(newFormula, newAssignment) {
		ans.mu.Lock()
		ans.assignment = newAssignment
		ans.satisfied = true
		ans.mu.Unlock()
		return
	}

	if !checkClauseValidity(formula) {
		return
	}

	selectedLiteral, err := selectLiteral(newFormula, newAssignment)
	if err != nil {
		return
	}

	assignment1 := maps.Clone(newAssignment)
	assignment1[selectedLiteral] = true

	simplifiedFormula := simplifyFormula(newFormula, selectedLiteral)

	wg.Add(1)
	go recursiveSolution(simplifiedFormula, assignment1, wg, ans)

	assignment2 := maps.Clone(newAssignment)
	assignment2[selectedLiteral] = false

	simplifiedFormula = simplifyFormula(newFormula, -selectedLiteral)

	wg.Add(1)
	go recursiveSolution(simplifiedFormula, assignment2, wg, ans)
}
